### PR TITLE
[WIP] ref: deprecated results obj, added support for simpler comms (1/n)

### PR DESCRIPTION
**Pull request by williamFalcon**
Here we move to deprecate results objects in favor of simpler syntax. Results obj support will remain until 0.10.0 but not 1.0.0.
(ie: 0.10.0 is 1.0.0 but backwards compatible for anyone who really needs it).

The new pattern is to decouple logging from hooks. That means that step, step_end, epoch_end are independent from logging.

## example
```python

def training_step(...):
    loss = ...
    return loss

# equivalent
def training_step(...):
    return {'loss': loss}

# to log
def training_step(...):
    self.log('anything', x, on_step=True, on_epoch=True)
```

## Passing around step results.
In the case the user still needs/wants to do something with the output of each batch, the other hooks are still there.

```python
def training_step(...):
    return {'loss': loss, 'random_thing': [1, 'a', Tensor(), ...]}

def training_epoch_end(self, training_step_outputs):
    for d in training_step_outputs:
        random_thing = d['random_thing']
    
```

## Hooks and .log are decoupled...
user can log from anywhere

```python
def training_epoch_end(...):
    some_new_val = ...
    self.log('my_new_val', some_new_val)
```

cc @ananthsub @awaelchli @justusschock

### CODE REVIEW

File: pytorch_lightning/core/step_result.py [hunk 1]
```diff
@@ -161,21 +161,22 @@ def log(
...
-                on_step,
-                on_epoch,
-                reduce_fx,
-                tbptt_reduce_fx=tbptt_reduce_fx,
-                tbptt_pad_token=tbptt_pad_token,
-            )
 
-            # set the value
-            self.__setitem__(name, value)
+        # always log the original metric
```
Comment (williamFalcon) [inline pytorch_lightning/core/step_result.py:165]: @edenafek  remember which ticket numbers this solved?

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -146,6 +149,78 @@ def forward(self, x):
...
+                value,
+                on_step=False,
+                on_epoch=False,
+                logger=True,
+                prog_bar=False,
+                reduce_fx=torch.mean,
+                enable_graph=False
+            )
+
+
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:187]:
```suggestion
```
you do not like long docstring and this has zero information value

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -146,6 +149,78 @@ def forward(self, x):
...
+                name,
+                value,
+                on_step=False,
+                on_epoch=False,
+                logger=True,
+                prog_bar=False,
+                reduce_fx=torch.mean,
+                enable_graph=False
+            )
+
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:186]:
```suggestion
```
you do not like long docstring and this has zero information value

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -146,6 +149,78 @@ def forward(self, x):
         if self.trainer.is_global_zero:
             print(*args, **kwargs)
 
+    def log(
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:152]:
```suggestion
    def track(
```
would it be better name which dos not colide with loggers...

File: pytorch_lightning/trainer/training_loop.py [hunk 5]
```diff
@@ -490,6 +564,13 @@ def run_training_epoch(self):
             if self.trainer.should_stop:
                 break
 
+            # stop epoch if we limited the number of training batches
+            if batch_idx + 1 >= self.trainer.num_training_batches:
+                break
+
+            # progress global step according to grads progress
+            self.increment_accumulated_grad_global_step()
+
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/training_loop.py:573]: @awaelchli  here.

File: pytorch_lightning/trainer/training_loop.py [hunk 6]
```diff
@@ -504,6 +585,10 @@ def run_training_epoch(self):
         # epoch end hook
         self.run_on_epoch_end_hook()
 
+        # increment the global step once
+        # progress global step according to grads progress
+        self.increment_accumulated_grad_global_step()
+
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/training_loop.py:591]: @awaelchli  and here

File: pytorch_lightning/core/lightning.py (hunk 3 above)
Comment (williamFalcon) [reply pytorch_lightning/core/lightning.py:186]: ? the docstrings are there... did you look?

Comment (williamFalcon) [reply pytorch_lightning/core/lightning.py:186]: image

Comment (Borda) [reply pytorch_lightning/core/lightning.py:186]: Yes and I recommend to remove the lines with default as it is already in function fingerprint

File: pytorch_lightning/trainer/logging.py [hunk 7]
```diff
@@ -57,6 +58,23 @@ def process_dict_result(self, output, train=False):
...
+            for k, v in output.items():
+                if k in ['log', 'progress_bar']:
+                    m = f"""
+                    The {{{k}:dict keyword}} was deprecated in 0.9.1 and will be removed in 1.0.0
+                    Please use self.log(...) inside the lightningModule instead.
+ 
+                    # log on a step or aggregate epoch metric to the logger and/or progress bar
+                    # (inside LightningModule)
+                    self.log('train_loss', loss, on_step=True, on_epoch=True, prog_bar=True)
+                    """
```
Comment (carmocca) [inline pytorch_lightning/trainer/logging.py:75]: Please, do not use triple quote strings for warnings and logs since they include all the extra whitespace:

```
>>> import warnings
>>> m1 = (
...     "this is "
...     "a "
...     "test"
... )
>>> warnings.warn(m1)
<stdin>:1: UserWarning: this is a test
>>> m2 = """
...   this
...   is a
...   test
... """
>>> warnings.warn(m2)
<stdin>:1: UserWarning: 
  this
  is a
  test

```

Comment (williamFalcon) [reply pytorch_lightning/trainer/logging.py:75]: i disagree with this... 

these long strings are more readable to the user when the get this kind of warning.

Comment (carmocca) [reply pytorch_lightning/trainer/logging.py:75]: The problem is not that it is multiline, but having the whitespace at the beginning depends on the indentation level.

For that case in particular:
```
<stdin>:1: UserWarning: 
                    The {{{k}:dict keyword}} was deprecated in 0.9.1 and will be removed in 1.0.0
                    Please use self.log(...) inside the lightningModule instead.
 
                    # log on a step or aggregate epoch metric to the logger and/or progress bar
                    # (inside LightningModule)
                    self.log('train_loss', loss, on_step=True, on_epoch=True, prog_bar=True)

```

You would have to either remove the indentation:
```python
                    m = f"""
The {{{k}:dict keyword}} was deprecated in 0.9.1 and will be removed in 1.0.0
Please use self.log(...) inside the lightningModule instead.

# log on a step or aggregate epoch metric to the logger and/or progress bar
# (inside LightningModule)
self.log('train_loss', loss, on_step=True, on_epoch=True, prog_bar=True)
                    """
```
(pretty ugly if you ask me)

or split it and print it with more than one warnings/logs

Comment (Borda) [reply pytorch_lightning/trainer/logging.py]: I agree with @carmocca, if you need an extra line, just use `\n`

**Comment by nazim1021** [comment]
cc @ananthsub @awaelchli @justusschock
Hi, after updating to version 1.0.4,  i think  below approach seems to be not working as desired

def training_step(...):
    return {'loss': loss, 'random_thing': [1, 'a', Tensor(), ...]}

def training_epoch_end(self, training_step_outputs):
    for d in training_step_outputs:
        random_thing = d['random_thing']
    
`training_step_outputs` is always empty. On further debugging i find that `training_step` is never getting called, instead this function `training_step_and_backward` is called with `run_training_batch`

**Comment by williamFalcon** [comment]
ok, will fix asap. thanks for reporting!

mind replicating it on a boring model?

@edenafek (p0)

**Comment by edenlightning** [comment]
@nazim1021! Thank you for reporting! I created an issue to track this, would be great if you can try to [repro using boring model](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and paste a colab link in the new issue](https://github.com/PyTorchLightning/pytorch-lightning/issues/4452)

**Comment by nazim1021** [comment]
@williamFalcon @edenlightning  sure, of course. Thanks for prompt reply
i ll add the colab link to new issue
